internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,351 @@
+/*
+Copyright 2021 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/ceph/ceph-csi/internal/util"
+
+	librbd "github.com/ceph/go-ceph/rbd"
+	"github.com/kube-storage/spec/lib/go/replication"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+//  imageMirroringMode is used to indicate the mirroring mode for an RBD image.
+type imageMirroringMode string
+
+const (
+	// imageMirrorModeJournal uses journaling to propagate RBD images between
+	// ceph clusters.
+	imageMirrorModeJournal imageMirroringMode = "journal"
+	// imageMirrorModeSnapshot uses snapshots to propagate RBD images between
+	// ceph clusters.
+	imageMirrorModeSnapshot imageMirroringMode = "snapshot"
+)
+
+// imageMirroingState represents the image mirroring state.
+type imageMirroingState string
+
+const (
+	// If the state is up+replaying, then mirroring is functioning properly.
+	// up means the rbd-mirror daemon is running, and replaying means
+	// this image is the target for replication from another storage cluster.
+	upAndReplaying imageMirroingState = "up+replaying"
+	// If the state is up+stopped means the rbd-mirror daemon is
+	// running and stopped means the image is not a target for replication from
+	// another cluster
+	upAndStopped imageMirroingState = "up+stopped"
+)
+
+// mirroringMode is the key to get the imageMirroringMode from parameters.
+var imageMirroringKey = "mirroringMode"
+
+// forceKey is the key to get the force option from parameters.
+var forceKey = "force"
+
+// getVolumeFromID gets the rbd image details from the volumeID.
+func (cs *ControllerServer) getVolumeFromID(ctx context.Context, volumeID string, secrets map[string]string) (*rbdVolume, *util.Credentials, error) {
```

Comment:
This looks like a function that should be part of `internal/rbd/controllerserver.go` and can be used there too.

Reply:
Yes can be moved later to do optimization in the rbd controllerserver.go

Reply:
please mark this with a `TODO:` comment

Reply:
done